# Week 1: Python Basics & Flow Control
AI for Good - Minor ADSAI

Work through the sections below in order. Write your code in the empty cells.
Run a cell with Shift + Enter. Getting errors is part of the method, not a
sign that something is wrong.

Today runs in **three learning waves**. Each wave starts with a short
mini-lecture, then you code. A wave starts at every horizontal line in this
notebook, marked like this:

**LEARNING WAVE 1 · SLIDES 5-8**

- Wave 1 - section 1: Python as a calculator
- Wave 2 - section 2: Variables and types
- Wave 3 - section 3: Making decisions
- Then section 4: the big individual assignment, and sections 5-7: wrap-up
  (GitHub portfolio, AI app builders, and what to remember)

Do not run ahead into the next wave - wait for the mini-lecture.

---

**LEARNING WAVE 1 · SLIDES 5-8, 30 minutes**

## 1. Python as a calculator

Sections 1.1 to 1.4. The arithmetic operators, printing results, and your first calculation with real climate data.

### 1.1 Arithmetic symbols

Python uses these symbols to calculate:

- Add: `+`
- Subtract: `-`
- Multiply: `*`
- Divide: `/`
- Integer division: `//`
- Modulo: `%`
- Exponent: `**`

Show the result of:
- 8 times 10
- 10 to the power of 3
- 300 minus 25 times 10 (check whether the normal order of operations is applied)

Use the 3 cells below for your answers.

In [ ]:
8 * 10

80

In [ ]:
10 ** 3

1000

In [ ]:
300 - 25 * 10

50

### 1.2 Printing results

A cell only shows the result of its *last* line. Use the print function to show
the results of all three calculations below in one cell:
- 8 times 10
- 10 to the power of 3
- 300 minus 25 times 10

In [ ]:
print(8 * 10)
print(10 ** 3)
print(300 - 25 * 10)#

80
1000
50


- What happens if you put a hashtag (#) in front of one of the print statements?

### 1.3 Working out the operators

*Slide 7*

Nobody has told you yet what `//`, `%` and `**` do. You are going to work that
out yourself.

- First write down your guess for each of the four lines below.
- Then run the cell and compare the output with your guess.

In [ ]:
print(7 / 2)
print(7 // 2)
print(7 % 2)
print(7 ** 2)#

3.5
3
1
49


- Run the cell below as well, and look at both outputs together.

In [ ]:
print(9 / 3)
print(9 // 3)
print(9 % 3)
print(11 % 3)
print(10 % 2)
print(11 % 2)

3.0
3
0
2
0
1


- Explain in your own words what each of these four operators does:
`/`, `//`, `%` and `**`.

Your answer:
Divide: /
Integer division: //
Modulo: % percenteges of something
Exponent: ** like 3 to the power of 9

> Add blockquote




- One of these operators answers a question programmers ask constantly:
*is this number even or odd?* Which one, and how? Show it in the cell below
for the numbers 10 and 11.

In [ ]:
print(10 % 2) # Expected output: 0 (even)
print(11 % 2) # Expected output: 1 (odd)

0
1


### 1.4 Climate arithmetic

*Slide 8 - 15 minutes*

A return flight from Amsterdam to New York emits about 1258 kg of CO2 per
passenger. This is real data - your code is working with real-world impact.

- Print the total CO2 for 3 passengers, using an expression. No variables yet:
just numbers and operators.

In [ ]:
print(1258 * 3)

3774


---

**LEARNING WAVE 2 · SLIDES 9-11, 40 minutes**

## 2. Variables and types

Sections 2.1 to 2.6. Storing values under a name, the four data types, converting between them, and asking the user a question.

### 2.1 Variables

- Define the following variables and assign values to them:

`flight_co2`:
the emission per passenger is 1258.

`passengers`:
the number of passengers is 3.

`total_co2`:
the total emission is the emission per passenger times the number of passengers.

- Use the variables to print the total.

In [ ]:
flight_co2 = 1258
passengers = 3
total_co2 = flight_co2 * passengers

print(total_co2)

3774


### 2.2 Naming conventions

- What are the most important conventions for variable names in Python?
- Why is it important to stick to them?

Your answer:


- Run the code in the cell below. The result is 1258, because that is the
highest number in the list.
- Now remove the hashtag in front of `max = 15` and run the cell again. What happens?

In [ ]:
highest = max([1258, 400, 230])
print(highest)

1258


- You now have to restart the kernel to make `max()` work again.

### 2.3 Types

- Apply the `type()` function to determine the data type of `flight_co2`,
`passengers` and `total_co2`. Print the results.

In [ ]:
print(type(flight_co2))
print(type(passengers))
print(type(total_co2))

<class 'int'>
<class 'int'>
<class 'int'>


- What is the difference between an integer and a float?

Your answer:


- Create variables with your name and whether you cycle to university.
For example:

`my_name = "Amara Okonkwo"`
`cycles = True`

- Print the types of both variables.

In [ ]:
my_name = "Dylan"
cycles = True

print(type(my_name))
print(type(cycles))


<class 'str'>
<class 'bool'>


- Explain what the following types are: string, integer, float and Boolean.

Your answer: string is a data type, integer is a whole number, float is a number with decimal point, boolean is like true a fault


- Sometimes numbers are stored as strings. The variable `number` below
contains the string `'1258'`.
- Convert it to an integer and print the type of the result.

In [ ]:
number = '1258'


### 2.4 Operations with strings

- Convert `number` back to a string.
- Create the variable `add_number` that contains `number` plus `number`.
- Print `add_number`.

In [ ]:
add_number = number + number
print(add_number)

12581258


You can see here that the `+` operator behaves differently for a string
than it does for an integer or a float.

### 2.5 Printing combinations of different data types

To print variables of different types with one print function, you have to
convert them all to strings first. You can then join them with `+`.

In section 2.1 you made the variables `passengers` and `total_co2`. They have
different data types.

- Print the following sentence using both variables:

`A group of 3 passengers emits 3774 kg CO2 on this flight.`

In [ ]:
print('A group of 3 emits ' + str(passengers) + str(total_co2) + ' kg CO2 on this flight.'  )

A group of 3 emits 33774 kg CO2 on this flight.


### 2.6 Talking to the user

*Slide 11 - 15 minutes*

Training a large AI model can emit as much CO2 as five cars over their entire
lifetime, and every chat with an AI model uses energy too. You are going to
estimate your own AI energy footprint.

- Ask the user for their name using `input()`.
- Ask how many AI chats they have per day, and convert the answer to an integer.
- Calculate `weekly_chats`: the daily number times 7.
- One chat uses about 0.003 kWh. Calculate `weekly_energy`.
- Print: `Hi [name]! Your weekly AI energy use is [X] kWh.`

In [1]:
name = input("What is your name? ")
daily_chats_str = input("How many AI chats do you have per day? ")
daily_chats = int(daily_chats_str)

weekly_chats = daily_chats * 7
energy_per_chat = 0.003
weekly_energy = weekly_chats * energy_per_chat

print(f"Hi {name}! Your weekly AI energy use is {weekly_energy} kWh.")

KeyboardInterrupt: Interrupted by user

- Which data type does `input()` always return? Why do you have to convert it?

Your answer:


---

**LEARNING WAVE 3 · SLIDES 12-15, 20 minutes**

## 3. Making decisions

Sections 3.1 to 3.4. Comparisons, Booleans, and the if / elif / else block that is the core of every rule-based decision system.

### 3.1 Comparison operators

Python uses these symbols to compare two values:

- Equal to: `==`
- Not equal to: `!=`
- Greater than: `>`
- Less than: `<`
- Greater than or equal to: `>=`
- Less than or equal to: `<=`

Every comparison gives back a Boolean: `True` or `False`.

- Use print to answer the questions below:
- Is 5 equal to 5?
- Is 5 equal to `'5'`?
- Is the total CO2 of 3 passengers greater than 3000?

In [ ]:
if 5==5
print(True)
if 5=='5'
print(False)
if total_co2 > 3000
print(True)

- Explain in your own words the difference between `=` and `==`. = is just like == but its confuese i think that why there is ==

Your answer:


### 3.2 Boolean operators

`and` gives True only when both sides are True. `or` gives True when at least
one side is True. `not` flips the value.

- Write down your guess for each line, then run the cell.

In [ ]:
print(True and False)
print(True or False)
print(not True)
print(1258 > 1000 and 3 > 5)

False
True
False
False


### 3.3 if / elif / else

*Slide 15 - 20 minutes*

Many AI systems make decisions with exactly the structure you are about to
write. The same logic that flags a "high-risk loan applicant" can reproduce
human bias.

Python checks the conditions from top to bottom and runs the first one that is
True. Indentation is not styling in Python - it is the syntax that decides
which lines belong to which branch.

- Create the variable `income` with `income = int(input('Annual income in euros: '))`
- Write an `if` / `elif` / `else` block that assigns a risk level:
  - `income` below 20000 gives `'high risk'`
  - `income` below 40000 gives `'medium risk'`
  - anything else gives `'low risk'`
- Print: `Based on income alone, this person is classified as: [risk level]`
- Then also ask for `age`, and combine the two with `and` or `or` so that the
  classification uses both.

In [2]:
income = int(input('Annual income in euros: '))
if income < 20000:
    print('high risk')
elif income < 40000:
    print('medium risk')
else:
    print('low risk')

Annual income in euros: 3000
high risk


### 3.4 Reflection

- Is it fair to classify people on income alone? Which information is missing
from this model, and who would be harmed if a real organisation used it?

Your answer:


---

**BIG INDIVIDUAL ASSIGNMENT · SLIDES 17-19**

## 4. Green or Not?

Everything from the three waves, combined into one program. Individual work. This one goes into your portfolio.

In this assignment you combine everything from today into one program: input,
type conversion, calculation, and a decision. This is a portfolio piece - polish
it at home if you do not finish it in class.

**Data:**

Return short-haul flight: 400 kg CO2 each
Driving: 0.23 kg CO2 per km
Meat-heavy diet: 1400 kg CO2 per year
Sustainable annual budget: 2500 kg CO2

**Classification thresholds:**

Below 2500 kg: under budget
2500 to 4000 kg: slightly over budget
Above 4000 kg: far over budget

**Assignment:**
- Make variables for the four reference values above.
- Ask the user how many return flights they take per year, how many kilometres
  they drive per year, and whether they eat meat daily. Convert each answer to
  the right type.
- Calculate the total annual CO2 and assign it to a variable.
- Use `if` / `elif` / `else` to classify the total against the thresholds.
- Print the type of one of your variables.
- Print a personalised message containing both the total and the classification,
  for example: `You emit 5150 kg CO2 per year. That is far over budget.`

In [3]:
# reference values:

# questions to the user:

# calculation:

# classification:

# print the result:


### 4.1 Bonus reflection

- Suppose a government used this calculator to decide who pays a climate tax.
What would have to be true about the model before that would be acceptable?

Your answer:


---

**WRAP-UP · SLIDES 21-24**

## 5. Set up your GitHub portfolio

Your entire grade for this minor is a portfolio, and this is where it lives.
Do this now, in class. Nobody leaves before their repository exists.

### 5.1 What GitHub actually is

Three words cover almost everything you need today:

- A **repository** (or *repo*) is a project folder that remembers its own history.
- A **commit** is a saved snapshot of that folder, with a short message saying
  what changed. It is a save point you can always go back to.
- **Push** means uploading your commits to GitHub so they exist somewhere other
  than your laptop.

That is it. You do not need the command line today - everything below happens in
the browser.

- Why is it useful that every commit carries a message? Think about opening your
  own portfolio in six months, or a teacher opening it during grading.

Your answer:


### 5.2 Create your portfolio from the template

Follow along. We move together and nobody gets left behind - put your hand up
the moment a screen does not look like the one on the projector.

1. Make a free account on <https://github.com> if you do not have one, and
   verify the email it sends you (check your spam folder).
2. Open the course template: <https://github.com/BTuyn/AI4G-Portfolio-Template>
3. Click the green **Use this template** button, then **Create a new repository**.
   Do *not* click *Fork* - it looks similar and behaves differently.
4. Name it `ai4g-portfolio-yourname`.
5. Click **Create repository**.

**Checkpoint - do not continue until this is true:** you are looking at a
repository under *your own* username, and it contains `Term 1/`, `Term 2/`,
`README.md` and `PORTFOLIO_CHECKLIST.md`.

### 5.3 Make it yours

The template ships with placeholder details. Replace them with your own.

1. Click `README.md` in your new repository.
2. Click the pencil icon to edit it in the browser.
3. Fill in your name, student number and cohort.
4. Scroll to the bottom. In the commit message box, type
   `Set up my portfolio` and click **Commit changes**.

You have just made your first commit.

### 5.4 Upload today's work

Your homework for every week goes in that week's `homework/` folder.

1. Navigate to `Term 1/Week 1/homework/`.
2. Click **Add file**, then **Upload files**.
3. Drag this notebook (`Week1_Workshop_Student.ipynb`) into the browser window.
4. Commit message: `Week 1 homework: Python basics and flow control`
5. Click **Commit changes**.

Refresh the page. Your notebook is now in your portfolio, with your name on it
and a timestamp. That is the whole workflow you will repeat every week.

### 5.5 Where everything goes, and the rules that matter

Every week folder has the same three subfolders:

| Folder | What goes in it |
|---|---|
| `homework/` | your workshop notebook and assignments |
| `hackathon/` | Friday's prototype code and slides |
| `presentation/` | slides and demo proof, in the weeks your group presents |

Four rules, and the first one is the one that actually gets people in trouble:

1. **Never commit API keys, tokens or passwords.** Once something is pushed it
   is in the history even if you delete it later. Put secrets in a `.env` file -
   the template already contains `.env.example` and a `.gitignore` that keeps
   `.env` out of your commits.
2. **Upload every week.** A portfolio committed entirely on the last night is
   extremely visible in the history, and it is not what is being assessed.
3. **No files over 100 MB.** Link to large datasets or videos instead.
4. **Say which AI tools you used** and what you changed yourself.

- Suppose you accidentally commit a file containing your API key, notice an hour
  later, and delete the file in a new commit. Is the key safe? What should you
  actually do?

Your answer:


### 5.6 Optional: doing this from your own machine

Everything above works in the browser, and the browser is enough for this
course. If you want to work locally later, the same four ideas have commands.
Install Git first, then in a terminal:

```bash
git clone https://github.com/YOUR-USERNAME/ai4g-portfolio-yourname.git
```

```bash
git add .
git commit -m "Week 1 homework: Python basics and flow control"
git push
```

`clone` downloads your repository once. `add` selects what to include, `commit`
takes the snapshot with its message, and `push` uploads it. That is the same
sequence you just did with buttons.

This is optional, and it is genuinely fine to finish the whole minor without
ever opening a terminal.

### 5.7 Before you leave

Check each of these yourself:

- [ ] A repository exists under my own GitHub username, made from the template.
- [ ] `README.md` contains my name, student number and cohort.
- [ ] Today's notebook is in `Term 1/Week 1/homework/`.
- [ ] My commit messages say what changed, not just "update".
- [ ] There are no keys, tokens or passwords anywhere in my repository.

Paste the link to your repository in the cell below so you have it to hand -
you will need it every single week.

My portfolio:


---

**WRAP-UP · SLIDES 25-28**

## 6. AI app builders: Lovable and bolt.new

Not a wave: the closing segment of the day, after the debrief.

In Friday's hackathon you will build something bigger than anything in this
notebook, and you will not write all of it by hand. Two tools let you describe an
app in plain English and get a working one back:

**Lovable** (<https://lovable.dev>) builds a full web app - screens, database,
login - from a description. It runs in the browser; you make a free account,
describe what you want, and refine it by chatting. You can publish it to a live
URL and sync the code to GitHub.

**bolt.new** (<https://bolt.new>) does the same kind of thing from StackBlitz.
It runs a real development environment inside your browser tab, so you see the
code, the installed packages and a live preview side by side. You can start
without an account on a free allowance, and you can push the result to GitHub.

Neither needs anything installed. Both give you real code that you own.

**The rule for this course:** read what the tool produces before you accept it,
and be able to explain every line you hand in. A prototype you cannot explain is
not a prototype you can present. You must also say which AI tools you used and
what you changed yourself - that is a portfolio requirement, not a formality.

- Use the cell below for your own notes during the demo.

Your notes:


## 7. Five things to remember

1. Python follows the normal order of operations: `2 + 3 * 4` is 14, not 20.
   Use brackets when in doubt.
2. `/` always gives a float, `//` throws away the remainder, and `%` gives you
   the remainder itself.
3. `input()` always returns a string. Convert it with `int()` or `float()`
   before you calculate with it.
4. Indentation is syntax, not style. Four spaces decide which lines belong to
   which branch.
5. Your `if` / `elif` / `else` today is the same structure behind AI classifiers
   tomorrow. Design the conditions carefully.

**Next week:** Loops & Functions

**Friday:** the hackathon. You find out on the day which tool you use and which
problem you solve.